parsl/dataflow/dflow.py

```diff
@@ -348,7 +348,7 @@ def handle_app_update(self, task_id, future, memo_cbk=False):
             for dfu in self.tasks[task_id]['app_fu'].outputs:
                 f = dfu.file_obj
                 if isinstance(f, File) and f.is_remote():
-                    f.stage_out(self.tasks[task_id]['executor'])
+                    self.data_manager.stage_out(f, self.tasks[task_id]['executor'])
```

Comment:
Does the data_manager at this point defer if the file is to be fetched by worker_side mechanism ?

Reply:
I don't think it is the role of `File` to understand how it gets moved between places. I think it is the role of DFK and DataManager to know that.

At the moment, they do it by launching a stage out parsl task.

I'd imagine that this line is approximately the point in the code where the data manager or DFK would choose the right way to do the staging, and make a different call if it isn't using the existing separate-task-staging - for example annotating the job to be run/wrapping it in some wrapper function

Reply:
stage_out will need to move *before* the task is submitted to the executor, though: once a task has finished, it's too late to start deciding on extra things to do inside the task.

In the existing separate-stage-out-task implementation, I think that can be made to work because these will be tasks dependent on some Future - although a bit fiddly.

Reply:
and for per-worker stage-ins, there does not need to be any DataFuture mediating the execution order between the in-worker stage in sub-task and the main sub-task - so can hopefully pass in a plain File object (with no `localpath` set yet) and have the stage in sub-task set `localpath` on the far end.

Reply:
Okay, this is consistent with my mental model where either the dfk/datamanager will have to do something like this at task launch time.

```pseudocode
If worker can access the file:
      do nothing
if worker can access file if brough to globus_ep:
      fetch with globus to dest ep via parsl side staging
if worker can access file only if brought to worker: # no shared-fs with ep
      fetch with globus to local ep via parsl side staging, if file is remote
      use worker side staging with task wrappers.
```

The `File` objects encode the information about different endpoints/shared-fs where they might have been fetched to.

We'll still need to expose methods to say move files to a particular site to user, ideally we shouldn't have to tell them about dfk/datamanager in order to do those.

Reply:
> We'll still need to expose methods to say move files to a particular site to user, ideally we shouldn't have to tell them about dfk/datamanager in order to do those.

 Is that you wanting users to be able to move files around in some way that isn't as an input/output of an App?

Reply:
Pretty much anything user-exposed needs a DFK: we already have mechanisms to avoid the user having to explicitly pass around a DFK (via `parsl.load`)

Reply:
Just thinking about weird cases where the user just wants to be able to move files and not do any apps. Or have the files move to multiple places.